code/go/0chain.net/smartcontract/storagesc/allocation.go

```diff
@@ -283,6 +283,59 @@ func (sc *StorageSmartContract) filterBlobbersByFreeSpace(now common.Timestamp,
 	})
 }
 
+// getBlobbers get blobbers from MPT concurrently based on input blobber ids
+func (sc *StorageSmartContract) getBlobbers(blobberIDs []string,
+	balances chainstate.StateContextI) (*StorageNodes, error) {
+
+	var (
+		err             error
+		errCount        int
+		poolSize        = 8
+		inputBlobbers   = new(StorageNodes)
+		inputBlobberMap = make(map[string]bool)
+	)
+
+	for _, b := range blobberIDs {
+		if _, ok := inputBlobberMap[b]; !ok {
+			inputBlobberMap[b] = true
+		}
+	}
+
+	input := make(chan string, len(inputBlobberMap))
+	output := make(chan *StorageNode, len(inputBlobberMap))
+	defer close(output)
+	for pool := 0; pool < poolSize; pool++ {
+		go func() {
```

Comment:
I don't think we should use concurrency in smart contracts tbh

Reply:
do we get any performance boost?

Reply:
My benchmarks suggest it can halve the speed of smart contracts. So using parallel for loops could well allow us to have a maximum of say 40 blobbers per allocation rather than 20.

What is the problem?

Reply:
40 blobbers, concurrency:8 get_blobbers	{"time_taken": "167ns"}
40 blobbers, concurrency:4 get_blobbers	{"time_taken": "208ns"}
40 blobbers, concurrency disabled get_blobbers	{"time_taken": "250ns"}

Reply:
these are timings for this function with and without cocurrency

Reply:
afaik we never used it and all structures used in smartcontracts don't have proper synchronization, if we allow to use it somewhere we should code accordingly, for example StateContextI is not designed for concurrent usage, look here [field access](https://github.com/0chain/0chain/blob/c73e2717354d70f33a33fe1a73487ca85f506d14/code/go/0chain.net/chaincore/chain/state/state_context.go#L290)

Reply:
in my opinion we should parallelize work on some higher level and implement smartcontracts just as not concurrent code